Missing recordings blog post (#2793)
PostHog/posthog.com: contents/blog/missing-recordings.md

featuredImage: ../images/blog/posthog-engineering-blog.png
featuredImageType: full
author: ["rick-marron"]
categories: ["Engineering", "Inside PostHog"]
---

[Session Recording](product/session-recording) is one of PostHog's core features. It allows you to see how real users use your website, mouse clicks and all. It's really cool and addictive (give it a try and you'll see what I mean).

Recently, we made a ton of improvements to the feature. To name a few, we:

* Decreased average recording load time from ~2.5s to 700ms
* Resolved a big bug that caused recordings to "wonk out" if a recording spanned multiple tabs or browser windows
* Revamped the player experience to make it faster for users to get to the moment in the recording they care about
* Improved the speed to query recordings, allowing users to search across all recordings at once (it used to be limited to a single day's worth of recordings)

But there was one fix that stood out in this quality push. Customers would often tell us recordings were missing, and we didn't know why.  

This is the story of how we decreased the number of missing recordings from 15% to <1%, and in doing so, moved the feature out of beta.

> >[Rick Marron](/handbook/company/team#rick-marron-software-engineer) is a software engineer at PostHog. In previous lives Rick was a Program Manager at Microsoft, and he spent over a year in Nairobi, Kenya where he developed the first publicly available real-time source for traffic data in the city using computer vision and street cameras. He's a native of Rochester, New York but lives in Los Angeles now because... sunshine. Follow him [on Twitter](https://twitter.com/rick_marron) and [Github](https://github.com/rcmarron). 

Reviewer: @rcmarron I've added this author bio (new thing I'm doing on the blog). Let me know if you're happy with it or change anything you don't like!

Author: Looks great to me!

# The mystery of missing recordings at PostHog

> Session recordings is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.

Reviewer 2: ```suggestion
Session Recordings is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.
```

---
date: 2021-06-10
title: The mystery of missing recordings at PostHog
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
featuredImage: ../images/blog/
featuredImageType: full
author: ["rick-marron"]
categories: ["Engineering"]
---

> Session Recording is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.

Reviewer: I think the intro should talk a little more about the Session Recording feature. How does it work, how do people use it etc. We want to talk up the feature as much as possible.

## Looking deeper

What was causing the rest of the missing recordings? With no more clues, it came time to dig through the mechanics of the code.

When a recording started, we took a 'full snapshot' of the page, and from that point forward, we would send up only the data that changed in the recording. For some reason, the missing recordings would often have the later data, but no full snapshot.

Why were the full snapshots missing? Presumably, if the later data was being sent up, the snapshot code should've already run. To further solidify this line of thinking, the determining the changes on the page fully depended on the snapshot existing in memory, so it must've already taken a full snapshot.

After a day or so of pounding our heads on the keyboard, we realized that recordings were being split when a user was inactive for 30 minutes. But when we split the recording, we were not sending up a new full snapshot.

We made [a fix](https://github.com/PostHog/posthog-js/pull/318) for this, and the missing recordings started to quickly drop.

## Slow moving metrics

> Because this fix went into posthog-js, and different customers upgrade at different speeds, it took a while to see the full effect of the fix in our metrics. However, after about a month of waiting, the 15% missing recordings metric dropped to <1%.

Reviewer: I feel like we're missing a conclusion here:

1) What did we learn from fixing this problem?
2) What feedback, if any, did we receieve from customers?

Author: 100% agree. Just added a conclusion that touched on the other quality improvements that we made, and the impact we've seen on usage.

---
date: 2021-06-10
title: The mystery of missing recordings at PostHog
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
featuredImage: ../images/blog/
featuredImageType: full
author: ["rick-marron"]
categories: ["Engineering"]
---

> [Session Recording](product/session-recording) is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.

Reviewer: Am I correct in thinking the feature was still in beta when this problem was highlighted? We should say that if it's the case.

Author: Good call - snuck in a line about moving it out of beta.

We decided to use the second metric because it would let us strive to 0% missing recordings. If you're curious, you can read the conversation that led to this decision in [this GitHub issue](https://github.com/PostHog/posthog/issues/5478).

In the first run of the metric, it showed that we we're missing ~15% of recordings across [PostHog Cloud](/pricing) users.

## Checking Sentry for clues

Now that we had a metric, we started hunting for clues. What was causing the missing recordings?

The first place we looked was Sentry. There were a couple of errors in the ingestion pipeline that seemed suspicious. The most notable was a "Max data upload size exceeded" error that was firing a few thousand times a day. This seemed like it could be an obvious cause of missing recordings. If some data was not being uploaded because it was too big, it would make sense that we could not get a 'full snapshot'.

To figure out what was causing this error, we looked at data that had been uploaded that was just below our size threshold of 20mb. It looked like large data URLs of images were the cause.

> To test this out, we made a filter that would remove data URLs from any payload that exceeded a 5mb threshold, and it would replace the image with a simple SVG.

Reviewer: Just looking at the PR, I think it would nice to add a little context over the SVG decision. There's a nice bit of back and forth and we can use the embedded images to add some visuals to the piece.

Author: Great idea. Pulled some of that discussion out into the post.

---
date: 2021-06-10

> title: The mystery of missing recordings at PostHog

Reviewer: ```suggestion
title: How we fixed a major issue with our Session Recording beta
```

Reviewer 3: totally subjective (feel free to ignore) flyby: I think the former sounds like it'll be a more interesting read

Reviewer: On reflection, you're totally the right

To figure out what was causing this error, we looked at data that had been uploaded that was just below our size threshold of 20mb. It looked like large data URLs of images were the cause.

To test this out, we made a filter that would remove data URLs from any payload that exceeded a 5mb threshold, and it would replace the image with an obvious placeholder.

![Recording Page with SVGs](https://user-images.githubusercontent.com/4813045/137223852-eeb4273e-d81c-49ca-a115-02b9cd0979e1.png)

One issue we faced was that we didn't want this feature to significantly increase the size of `posthog-js`, but we would have to include the placeholder image in the bundle. After some discussion, we landed on the idea of using a very small SVG that could easily scale to the size of the image being replaced. You can read more about the discussion that landed us here in [the PR itself](https://github.com/PostHog/posthog-js/pull/317).

> After deploying the fix, the Sentry error went away, but the missing recording only went down a few percent.

Reviewer: ```suggestion
After deploying the fix, two things happened:

1. The Sentry error went away 🥳
2. The missing recording only went down a few percent 😩

Back to the drawing board.
```

Author: Love the emojis!

## Looking deeper

What was causing the rest of the missing recordings? With no more clues, it came time to dig through the mechanics of the code.

When a recording started, we took a 'full snapshot' of the page, and from that point forward, we would send up only the data that changed in the recording. For some reason, the missing recordings would often have the later data, but no full snapshot.

Why were the full snapshots missing? Presumably, if the later data was being sent up, the snapshot code must've already run. To further solidify this line of thinking, the code to determine the changes to the page depends entirely on the snapshot existing in memory, so the system must've already taken a full snapshot.

> After a day or so of pounding our heads on the keyboard, we realized that recordings were being split when a user was inactive for 30 minutes. But when the split happened, we were not taking a new full snapshot.

Reviewer: I feel like there's some missing detail here that would really round off the narrative. What steps / dead ends did we / you go down while you were pounding heads on keyboards?

---
date: 2021-06-10

> title: The mystery of missing recordings at PostHog

Reviewer: ```suggestion
title: The mystery of missing session recordings at PostHog
```
